PowerNSX.psm1

```diff
@@ -21191,6 +21191,432 @@ function Remove-NsxSecurityGroupMember {
+            Add-XmlElement -xmlRoot $xmlDynamicCriteria -xmlElementName "value" -xmlElementText $value
+        }
+
+        $xmlDynamicCriteria
+
+    }
+
+    end{}
+}
+
+function Add-NsxDymanicMemberSet {
+
+    <#
+    .SYNOPSIS
+    Adds a new dynamic member set to an existing NSX Security Group.
+
+    .DESCRIPTION
+    An NSX Security Group is a grouping construct that provides a powerful
+    grouping function that can be used in DFW Firewall Rules and the NSX
+    Service Composer.
+
+    This cmdlet adds a new dynamic member set to an existing NSX Security Group.
+
+    A Security Group can consist of Static Includes and Excludes as well as
+    dynamic matching properties.
+
+    A dynamic member set defines the criteria that an object must meet for it to
+    be added to the security group. This gives you the ability to include
+    virtual machines by defining a filter criteria with a number of parameters
+    supported to match the search criteria.
+
+    A Security Group can have multiple dynamic member sets in an logical AND/OR
+    arrangement, and within each of the dynamic member sets, the set may contain
+    multiple dynamic criteria in an AND/OR arrangement.
+
+    .EXAMPLE
+
+    $criteria1Spec = New-NsxDynamicCriteriaSpec -key VM.name -operator AND -criteria contains -value "PROD"
+
+    $criteria2Spec = New-NsxDynamicCriteriaSpec -key VM.GUEST_OS_FULL_NAME -operator AND -criteria contains -value "Win"
+
+    $sg1 = New-NsxSecurityGroup -Name "SG-Production-Windows"
+
+    Get-NsxSecurityGroup "SG-Production-Windows" | Add-NsxDymanicMemberSet -operator OR -DynamicCriteriaSpec $criteria1Spec,$criteria2Spec
+
+    .EXAMPLE
+
+    $criteria3Spec = New-NsxDynamicCriteriaSpec -key VM.SECURITY_TAG -operator AND -criteria starts_with -value "ST_PCI"
+
+    $criteria4Spec = New-NsxDynamicCriteriaSpec -entity $(Get-Cluster DMZ) -operator AND
+
+    $sg2 = New-NsxSecurityGroup -Name "SG-DMZ-PCI"
+
+    Get-NsxSecurityGroup "SG-DMZ-PCI" | Add-NsxDymanicMemberSet -operator AND -DynamicCriteriaSpec $criteria3Spec,$criteria4Spec
+
+    .EXAMPLE
+
+    $criteria5Spec = New-NsxDynamicCriteriaSpec -key VM.SECURITY_TAG -operator AND -criteria starts_with -value "ST_Backup"
+
+    $criteria6Spec = New-NsxDynamicCriteriaSpec -entity $(Get-Cluster Dev-CL-01) -operator AND
+
+    $criteria7Spec = New-NsxDynamicCriteriaSpec -entity $(Get-NsxLogicalSwitch LS-Backup-Net) -operator AND
+
+    $criteria8Spec = New-NsxDynamicCriteriaSpec -key VM.NAME -operator AND -criteria contains -value "PROD"
+
+    $sg3 = New-NsxSecurityGroup -Name "SG-Backup-Clients"
+
+    $sg3.objectid | Add-NsxDymanicMemberSet -operator OR -DynamicCriteriaSpec $criteria5Spec,$criteria6Spec
+    $sg3.objectid | Add-NsxDymanicMemberSet -operator OR -DynamicCriteriaSpec $criteria7Spec,$criteria8Spec
+
+    #>
+
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidDefaultValueSwitchParameter","")] # Cant remove without breaking backward compatibility
```

Comment:
I'll submit PR for firewall rule on edge shortly - i have a new way of handling confirm moving forward that will deprecate the -confirm switch.  For now the way you have it currently is required, but in very near future i want to change to visibly deprecate the -confirm:$false approach in favour of a -NoConfirm switch (will remove need for this decoration, and comply with bp in powershell code analyser tool.  Just FYI for moment...